### PR TITLE
API update proposal

### PR DESCRIPTION
**Pull request by benogle**
WIP 

I've been thinking about the API a lot, writing providers, and talking to @nathansobo about what we want in terms of API. Generally these proposed changes are naming related and improvements for clarity, consistency and ease of use.

Here are all options in an example provider. I'll explain each bit in some detail below.

```coffee
class SomeProvider
  selector: '.source.js, .source.coffee'
  disableForSelector: '.source.js .comment'
  inclusionPriority: 1
  excludeLowerPriority: true

  getSuggestions: ({editor, position, scopeDescriptor, prefix}) ->
    new Promse (resolve) ->
      suggestion =
        text: 'someText' # OR
        snippet: 'someText(${1:myArg})'
        replacementPrefix: 'so' # (optional)
        leftLabel: 'NSView *' # (optional)
        rightLabel: '' # (optional)
        leftLabelHTML: 'NSView *' # (optional)
        rightLabelHTML: '' # (optional)
        type: 'function' # (optional)
        # icon: '' # (optional)
      resolve([suggestion])

  # (optional): called _after_ the suggestion `replacementPrefix` is replaced by the
  # suggestion `text` in the buffer
  onDidInsertSuggestion: ({editor, triggerPosition, suggestion}) ->

module.exports =
  activate: ->
    atom.services.provide('autocomplete.provider', '2.0.0', new SomeProvider)
```

## No id

It seems like the only reason we need the `id` would be for the provider blacklist construct. I am proposing we remove that construct, so this `id` is no longer necessary.

## blacklist -> disableForSelector

Renamed for clarity

## Priority system

In place of the provider blacklist, we thought a priority system might work better. This will still allow multiple providers to give suggestions, and allow for another main provider to disable the default.

Responses from would be sorted by `priority` high to low. The `allowLowerPriority` property would allow a provider to disable those lower than itself. We might have conventions for priority: accessory providers like the snippets are 100, default is 0, and generally 'main' providers like clang are 1.

Additionally this is consistent with our other APIs like panels, the status bar, and soon-to-be gutters.

```coffee
class SnippetsProvider
  priority: 100

class PathsProvider
  priority: 100

class ClangProvider
  priority: 1
  allowLowerPriority: false
  allowLowerPriorityWhenEmpty: true # maybe later

class DefaultProvider
  priority: 0
```

## Changes to requestHandler()

* Rename to `getSuggestions` as it is a much better description of the method action.
* Remove extraneous parameters.
  * Previously: `{editor, buffer, cursor, position, scope, scopeChain, prefix}`
  * Now: `{editor, position, scopeDescriptor, triggerPrefix}`
    * `buffer` and `cursor` can be had from `editor`
    * `scope` -> `scopeDescriptor` for clarity, `scopeChain` should not be necessary. If they need a scope list, they can use `scopeDescriptor::getScopesArray()`
    * `prefix` -> `triggerPrefix`

### Prefixing changes

I want it to search until the previous word break. Only a small change. Here are some would-be test cases:

* `self ` -> `' '`
* `self::` -> `'::'`
* `self.` -> `'.'`
* `sel` -> `'sel'`
* `some-sel` -> `'some-sel'`

## Suggestion object changes

```coffee
suggestion =
  text: 'someText' # OR
  snippet: 'someText(${1:myArg})'
  replacementPrefix: 'so' # (optional)
  leftLabel: 'NSView *' # (optional)
  rightLabel: '' # (optional)
  type: 'function' # (optional)
  # icon: '' # (optional)
```

* `text` renamed from `word`
* `snippet` String indicating snippet text. Will prefer this over `text`
* `replacementPrefix` renamed from `prefix` as I want to be clear that this prefix will be replaced. When not specified, it the `triggerPrefix` will be used as the `replacePrefix`
* `leftLabel` HTML string. Displays to the left of the match. If this exists, always set as html
* `rightLabel` HTML string. Same as `label`. If this exists, always set as html
* `type` A string indicating the type of suggestion: 'function', 'function', 'variable', 'class', 'snippet', mixin', ...
  * Allows us to display an icon for the suggestion type
* `icon` This could be a path to an image to replace the default icon generated based on the `type`

I want to use these last four optional arguments to setup a UI similar to other editors. e.g. method

## onDidInsertSuggestion

Rather than attaching a callback to each suggestion, I think we should have a method on the provider to accept the callback:

```coffee
onDidInsertSuggestion: ({editor, triggerPosition, suggestion}) ->
```

`onWillInsert` has been removed as it looked like it was not used anywhere.

## Provide the provider or a list of providers

Finally, providing your provider is awkward today

```coffee
atom.services.provide('autocomplete.provider', '2.0.0', {provider: new SomeProvider})
```

Instead, we should just provide the provider or list of them:

```coffee
atom.services.provide('autocomplete.provider', '2.0.0', new SomeProvider)

providers = [new SomeProvider, new OtherProvider]
atom.services.provide('autocomplete.provider', '2.0.0', providers)
```

## Conclusion

I defintely think @joefitzgerald , @park9140, and friends did a nice job with the structure of things. So thanks for getting this started.

This is obviously not fully implemented.

### CODE REVIEW

**Comment by park9140** [comment]
@benogle,

I like the large majority of these suggestions.  One thing, is there any reason why replacement prefix needs to be sent with a sugestion?  It seems like a good default is to just take the predetermined prefix as the default when unset.

**Comment by benogle** [comment]
> is there any reason why replacement prefix needs to be sent with a sugestion? It seems like a good default is to just take the predetermined prefix as the default when unset.

Yeah, this is a good idea. It currently is required. I've updated the main spec to the prefix is optional.

**Comment by joefitzgerald** [comment]
> It seems like the only reason we need the id would be for the provider blacklist construct.

Yep, you're correct - that's the only reason the `id` exists: to allow providers to flag that another provider shouldn't be run. Currently, it's not actually used - except for `FuzzyProvider`.

> blacklist -> disableForSelector

:+1: 

> Rename to `requestHandler` > `getSuggestions`

:+1: Although I like the semantics of seeing this as a request / response flow, as it implies the possibility of cancellation and timeout (which I hope we do ultimately add).

> Remove extraneous parameters

:+1:, except `prefix` > `triggerPrefix`. The prefix doesn't actually trigger anything. It might be the result of a trigger, but this seems to leak implementation details. 

I've actually been thinking about also including the function we use to determine the prefix (`@prefixForCursor(cursor)` - this would need to be evolved to take an options object), as some providers need to use a different regexp; this would allow them to do so, without doing this (potentially) expensive operation once per provider per request. 

Further, `cursor.getBeginningOfCurrentWordBufferPosition()` has a horrible default behavior: it searches back to the last line which is entirely whitespace. In a massive document of 1000's of lines with no empty lines, this has dire performance implications, and we need to ensure it is never part of the default prefix behavior in `autocomplete-plus`. I worked around this in `autocomplete-paths` by reimplementing the function there, but we should seriously ask the question: should we ever calculate the prefix in the request? I've landed on yes, as this may be done `n` times where there are `n` providers – so if we can do this in a reusable way that is generally applicable to most providers – we should.

> I want it to search until the previous word break.

This may result in many providers having to write relatively complex strip behavior or append portions of the prefix to their suggestions. Perhaps we should just have two prefixes - one that has the new behavior you desire, one that has the current behavior; ideally both can be computed via one look at the buffer.

> Suggestion object changes

:+1:, except `leftLabel` and `rightLabel`. What if the semantics of these change in a RTL language? I'd prefer to prefix `Label` with something denoting context or intent rather than position, allowing re-styling without an awkward opposites game.

> Rather than attaching a callback to each suggestion, I think we should have a method on the provider to accept the callback

:+1:

> Provide the provider or a list of providers

We use an object with a `provider` or `providers` key today to allow for easy extension of the API with unforeseen use cases in the future. I agree that – in the current state – it's potentially a needless circumlocution. 

## Priority System

I'm not sold on priorities as it could cause a weird race by authors toward the highest priority. I love that we currently prioritize providers and their suggestions by the specificity of the scopes with which they are registered - but that provides challenges for utility providers like autocomplete-snippets. Should snippets suggestions be above or below a specific language providers' suggestions? I'd argue lower, but others might argue higher. 

This implies a need for configuration, and a need to sensibly apply defaults. 

The blacklisting scenarios complicate things somewhat.

* Should a provider who is more specifically registered (as assessed by the current scope descriptor) be able to blacklist all providers that are less specifically registered? I'd argue: `no` as this impedes potentially novel providers that the provider author has not foreseen
* Should a provider who is more specifically registered (as assessed by the current scope descriptor) be able to blacklist specific provider(s) that are less specifically registered? I'd argue: `yes` as this allows a provider to prevent known-not-to-work-together providers for the current scope descriptor

The latter scenario implies that a provider might need to know some metadata about another provider (e.g. id or package name - but what about scenarios when a package has multiple providers, like `autocomplete-plus`, for example; there are others).

I think we have more work to do on the priority system proposal you have articulated.

## General Thoughts

This is great, we want to help. Would you like assistance implementing the suggestions above, or do you want to run with it yourself?

**Comment by joefitzgerald** [comment]
> is there any reason why replacement prefix needs to be sent with a sugestion? It seems like a good default is to just take the predetermined prefix as the default when unset.

:+1:

**Comment by joefitzgerald** [comment]
Hey @benogle would love to have a quick chat in https://gitter.im/atom-community/autocomplete-plus on this if you're open to it...

**Comment by nathansobo** [comment]
> is there any reason why replacement prefix needs to be sent with a sugestion? It seems like a good default is to just take the predetermined prefix as the default when unset.

Also :+1:

**Comment by benogle** [comment]
> Although I like the semantics of seeing this as a request / response flow, as it implies the possibility of cancellation and timeout (which I hope we do ultimately add).

I think if we had a cancel, we could have the provider return something with a `dispose` method. This would fit in with our current methodology in core.

> except prefix > triggerPrefix. The prefix doesn't actually trigger anything. It might be the result of a trigger, but this seems to leak implementation details.

Ok, we can leave it as `prefix`. My main goal was to reduce confusion between the `prefix` and the `replacementPrefix`. As in "why do I need to pass the prefix back when it supplies me the prefix." I think between the difference in naming `replacementPrefix` and the fact that it is optional helps alleviate the confusion.

> I've actually been thinking about also including the function we use to determine the prefix, as some providers need to use a different regexp

Providing a default prefix in `getSuggestions` is nice in the 80% case. Snippets, the default provider, and likely others can just use what's provided.

I think we should either provide a default prefix in `getSuggestions`, or not provide one or care about prefix at all. If we wanted to push the onus down to each provider to figure out its own prefix, they could do it on their own in `getSuggestions`. We wouldnt need a special method they provide.
For example, my [clang provider](https://github.com/benogle/autocomplete-clang/blob/53e5eb7833c8ce0927c0acc456469952a7aae0ee/lib/clang-provider.coffee#L27) has weird requirements, and wouldnt benefit from the passing in a prefix to make the decision to trigger.

> This may result in many providers having to write relatively complex strip behavior or append portions of the prefix to their suggestions.

I disagree. There will either be word characters or non-word chars. This already happens today, and it's fine because the non-word chars dont produce matches. I agree it might be a little strange to have either or. As an alternative we could split it up into multiple `prefix` vars. Like `wordPrefix` and `nonWordPrefix`. But I'm not super into it. It might look like:

```coffee
'self ': wordPrefix: '', nonWordPrefix: ' '
'self::': wordPrefix: '', nonWordPrefix: '::'
'self.': wordPrefix: '', nonWordPrefix: '.'
'self.method(': wordPrefix: '', nonWordPrefix: '('
'sel': wordPrefix: 'sel', nonWordPrefix: ''
'some-sel': wordPrefix: 'some-sel', nonWordPrefix: ''
```

> except leftLabel and rightLabel. What if the semantics of these change in a RTL language? I'd prefer to prefix Label with something denoting context or intent rather than position, allowing re-styling without an awkward opposites game.

RTL languages dont swap the entire concept of left and right, just the direction of writing.

Right and left labels are the most flexible and obvious way to specify what goes where. I do like the idea of semantic naming i.e. `returnType` or `objectType` etc. But it's considerably less flexible.

After talking to nathan about this, we feel that `leftLabel` and `rightLabel` should be rendered as text, and we should add `leftLabelHTML` and `rightLabelHTML` to optionally render as html.

> We use an object with a provider or providers key today to allow for easy extension of the API with unforeseen use cases in the future.

We're returning an object, though. If we wanted to support more, we could just add more to the provider object. Or if we wanted to provide something else, we could provide another object like `autocomplete.provider.something`. Since the service name is `autocomplete.provider`, I think we're pretty safe just returning the provider.

### Priority system

We talked about this and agree that conflating the inclusion and sorting with one priority is not good. So let's keep the sorting based on specificity. What if we renamed `inclusionPriority` and only used it for deciding inclusion in the suggestion list? An entire example:

```coffee
class SnippetsProvider
  selector: '*'
  inclusionPriority: 2

class PathsProvider
  selector: '*'
  inclusionPriority: 2

class CommentClangProvider
  selector: '.source.c .comment'
  inclusionPriority: 2
  excludeLowerPriority: true
  excludeLowerPriorityWhenEmpty: false

class ClangProvider
  selector: '.source.c'
  inclusionPriority: 1
  excludeLowerPriority: true
  excludeLowerPriorityWhenEmpty: true

class DefaultProvider
  selector: '*'
  priority: 0
```

As you mentioned we cannot use straight up specificity to disable the lower specificity things due to `SnippetsProvider` and `PathsProvider` being super low specificity. And we are super against explicit naming of providers. The best option we have is a priority system (another option is to `!important` accessory providers like `SnippetsProvider`. But ick).

Note the `CommentClangProvider` case. Pretend it has special suggestions while in `c` comments, and wants to disable the clang provider in that case. The solution is relatively elegant with the priority system.

### Help

Yes. Help would be nice. Either in other PRs for specific bits or as commits to this branch.

One thing that feels like it might be sticky is getting the deprecations right and upgrading data from providers using the old API. Do we pass the version of registration along with the provider registration? Then if they specify `2.0.0` we'd strictly use what they pass back, and if 1.0, convert to the 2.0 api, and emit deprecation messages.

**Comment by benogle** [comment]
I've finished the upgrading bits, and it supports everything laid out aside from the priority support. There are a couple paths now:

Dev provides with 1.0 or 1.1 api

1. Registration uses either 1.0 or 2.0 attributes on provider; no deprecations
1. User triggers ac+
1. All suggestions are converted to the 2.0 API (aside from hooks)

Dev provides with 2.0 API

1. 2.0 API used throughout; no deprecations

Dev provides old provider as 2.0 API

1. Deprecations provided for provider attributes on registration. Will not register without a `getSuggestions` method

Dev provides provider as 2.0 API that returns old suggestions

1. Deprecations provided the first suggestion when suggestions are returned; suggestions converted to the 2.0 API

Next step: the priority system.

**Comment by benogle** [comment]
This is ready to go. Doing a final pass through the diff and fixing anything that jumps out. Then merging

**Comment by joefitzgerald** [comment]
Reviewing now...

File: lib/autocomplete-manager.coffee [hunk 1]
```diff
@@ -110,26 +113,59 @@ class AutocompleteManager
...
-    cursorPosition = cursor.getBufferPosition()
-    currentScope = cursor.getScopeDescriptor()
-    return unless currentScope?
-    currentScopeChain = currentScope.getScopeChain()
-    return unless currentScopeChain?
-
-    options =
-      editor: @editor
-      buffer: @buffer
-      cursor: cursor
```
Comment (joefitzgerald) [inline lib/autocomplete-manager.coffee]: If we do ultimately extend our function `@prefixForCursor(cursor)` to the provider, it would best be done in this function and the cursor would be required. Should we leave the cursor in?

Comment (benogle) [inline lib/autocomplete-manager.coffee]: If we did ever do that we could pass the cursor to the provider's `getPrefix` method, no? Or we could pass the position. I really dont see the need for this method, though.

Comment (joefitzgerald) [inline lib/autocomplete-manager.coffee]: Many providers in the wild have a need to determine a different prefix. This is almost uniformly done via a slightly different regex. If we were to provide the function to them, they wouldn't all have to reimplement that - this is an area with potentially large performance implications. The default function we have today is susceptible to it also. 

See https://github.com/atom-community/autocomplete-paths/commit/8bd05ed5724683eaed798e5b8c49d99b29ec68a2#diff-7d05e6fe4c61ffced4fb4bd806573aadR23 through https://github.com/atom-community/autocomplete-paths/commit/8bd05ed5724683eaed798e5b8c49d99b29ec68a2#diff-7d05e6fe4c61ffced4fb4bd806573aadR51.

The issue is that in large files with no line breaks, `cursor.getBeginningOfCurrentWordBufferPosition({wordRegex: @wordRegex})` will actually search all the way up to the last non-blank row (which is always line 0). This causes the regexp to perform like crap, and consequently, typing is noticeably affected.

Offending `cursor` code: https://github.com/atom/atom/blob/a3f64ddf7afdebfded53777638214e92b9f5937a/src/cursor.coffee#L505-L506

Comment (joefitzgerald) [inline lib/autocomplete-manager.coffee]: Said another way, we should really provide a function to all providers (and use it in core) that only searches the current line for the prefix.

**Comment by joefitzgerald** [comment]
Before I forget to mention it, we should ensure https://github.com/atom-community/autocomplete-plus/wiki/Provider-API is updated with the semantics of v2.0.0. Longer term, we should determine how to migrate that content into `https://atom.io/docs/...`

File: lib/autocomplete-manager.coffee (hunk 1 above)
Comment (benogle) [inline lib/autocomplete-manager.coffee]: For now, they can do something simple like this. No need to reimplement that method

```coffee
regex = /[\w]+$/
line = editor.getTextInRange([[position.row, 0], position])
line.match(regex)?[0]
```

How should we provide that method to the provider? I dont think we should bake a prefix method into the cursor. The method being used has a different purpose. If they really really need the cursor, they can get it via `editor.getLastCursor()`. 

The parameters being passed provide parity with the old API. We can add more things later.

Comment (benogle) [inline lib/autocomplete-manager.coffee]: That could be even simpler as we are pretty permissive with the prefix:

```coffee
regex = /[\w]+$/
prefix.match(regex)?[0]
```

**Comment by benogle** [comment]
Yeah, I can update the wiki. Was planning on it after merging. As for the longer term thing, for now it's ok in the readme / wiki. We need to do this for a bunch of packages.

File: lib/autocomplete-manager.coffee (hunk 1 above)
Comment (joefitzgerald) [inline lib/autocomplete-manager.coffee]: I'm not suggesting the cursor has a function added to it. I'm suggesting we could additionally provide a getPrefix function on the options object which takes a cursor, and an options object. This is almost better illustrated via a commit that we can revert if needed. I'll have time later to do this.
  
```

Comment (benogle) [inline lib/autocomplete-manager.coffee]: A new PR would be nice. Please dont push that commit to this PR. 

I dont see why it would need the cursor. The cursor is a special purpose object. We only need the editor and position to generate a prefix. 

I am not fond of providing a random function in the options. It feels pretty dirty.

Comment (joefitzgerald) [inline lib/autocomplete-manager.coffee]: The old provider base class provided a getPrefix function. We ripped it out completely.

Comment (joefitzgerald) [inline lib/autocomplete-manager.coffee]: https://github.com/atom-community/autocomplete-plus/commit/690f7cade77ef3730b1248550a4e1503cee53db7#diff-43f9c4cec188d77db17033d59fb3944dL50

Comment (benogle) [inline lib/autocomplete-manager.coffee]: I think pulling it out was the right choice.

**Comment by lierdakil** [comment]
I've just noticed this, and I have a reservation about
> Remove extraneous parameters.
> Previously: {editor, buffer, cursor, position, scope, scopeChain, prefix}
> Now: {editor, position, scopeDescriptor, triggerPrefix}

I understand that in case of multiple cursors, autocompletion is only triggered for last added cursor, but it would be quite unintuitive to get that from `editor`. Meanwhile, position can be retrieved from `cursor`, so I would prefer to have `{editor, cursor, scopeDescriptor, prefix}` instead.

**Comment by nathansobo** [comment]
@lierdakil Your thoughts on `cursor` being provided make sense to me. @benogle what do you think?

**Comment by park9140** [comment]
@benogle, @lierdakil, @nathansobo,  I like the reduction of scope that we are passing to be just necessary components.  If we really want cursor we can get it from editor.  Position is part of the business logic of the system.  I would rather be exposing the results of what we have already done, and only providing the minimal of interfaces necessary to do anything else.  

That said I've been thinking about it, and I think we should really be limiting this further.  Editor is basically an unnecessary mass of information.  We should really only be exposing the text buffer and file name information since autocomplete providers shouldn't really be doing anything outside of looking at the text and saying, what should be completed here. 

If we really need to provide them access to change what the prefix is or get more information we should really provide other interfaces for enhancing that portion of the business logic.

**Comment by lierdakil** [comment]
@park9140 I **strongly** disagree. See https://github.com/lierdakil/autocomplete-haskell for example where `editor` is an integral part of logic.
Point is, some providers require information about editor instance, either for speed or convenience. I don't think artificially limiting provider interface is in any way a good idea. It's like closing a door and making guests enter through window. Possible, sure, but should you?
Passing reference to editor is in no way costly or confusing, and I see no reason not to, except for arbitrary "cleanliness".

As for needing cursor, Atom API provides some convenience methods on it. Most of information could be inferred by working directly on position and buffer, but there is no reason to reimplement something already in place, is there?

TL;DR I don't think that making provider developers jump through hoops is a good way to move forward. Artificially limiting or complicating API is just that.

**Comment by park9140** [comment]
@lierdakil,  I can see where you are coming from.  

However, the `autocomplete-haskell` plugin, can easily get the editors and do the work you are doing, by just calling the standard apis for getting editors.  By front loading this possibly expensive process you ensure that the plugin needs to do more work for any initial autocomplete request.

We should be focusing on responsiveness for autocomplete and anything that requires heavy analysis and binding should be heavily discouraged when handling autocomplete requests.  

In `atom-linter` we have an issue where the initial api is so wide open that every plugin ranges from near instant response to multiple second responses.  Most of this is because of bad code encouraged by providing a massive set of features on the primary api.  As a result the perception of the core plugin is that `atom-linter` is slow and that drives users away even though there are no other options.

Additionally, Adding the editors from here just makes the api more fragile.  By introducing the editor, or cursor here we take on issues with providers that exist solely because of changes to those two apis.  In my mind any plugin should use as little information provided here as possible.  If a plugin needs to do more it can acquire that information directly from the appropriate already existing apis.  These public apis will themselves go through proper deprication and update processes rather than us having to propagate any changes through the autocomplete api.

**Comment by joefitzgerald** [comment]
I also disagree @park9140. The grammar of the editor could be useful, the cursor could be useful, there are a number of things that the editor provides that are not available from the buffer, and there's no upwards pathway from the buffer to the editor. The converse is true.

I see where you're going, but removing the editor is too much IMO.

**Comment by lierdakil** [comment]
@park9140 you might not have noticed, but in `autocomplete-haskell`, I use reference to editor instance to access per-editor cache (which is initialized independently of autocompletion requests elsewhere), in order to speed up suggestion building. If editor reference is not passed to `requestHandler`, then I have no simple way of determining which cache to use, and would need to build suggestions from ground-up, or search all open editors for one I want. I can't stress enough that **your whole argument about responsiveness is completely backwards.**

Editor instance also provides things like `scanInBufferRange` and `backwardsScanInBufferRange`, which are immensely helpful in cases where built-in prefix detection does not cut it. And let's face it, every case is different, so there can be no "silver bullet" for built-in prefix detection.

As for discussion about Atom API vs ac-plus API, as far as I understand, general plan involves trying to get ac-plus into core packages, making this a moot point. Besides, simply exposing parts of Atom API via ac-plus API is not as fragile as you make it out to be. In any case, if Atom API introduces breaking changes, third-party providers *will* unavoidably break as well, regardless of if you provided editor instance or not.

So I stand my ground. I want editor and cursor references available in `requestHandler` (or `getSuggestions`, if you prefer) arguments. `editor` because it's simply easier and more effective, and `cursor` because in case of multiple cursors, it's not immediately obvious which one generated autocompletion request. `position` can be inferred from `cursor` with one method call, making it somewhat redundant.

**Comment by park9140** [comment]
@lierdakil, I saw the caching methodology you are using, and if we are providing enough information to tell what file you are autocompleting it is a moot point to tell where you should be caching.  Additionally, caching per editor is absolutely the wrong thing to do since multiple editors can be opened for the same file, so in your case caching based on the provided buffer would be more efficient.

As for prefix detection, I am totally on board with providing configurable prefix detection, which should honestly probably be provided the editor as an input.  However, I am against doing any recomputation of the prefix as a part of the api for requests.  We should really just provide a way for providers to set their prefix provider rather than change it mid request.

TLDR; It may be premature at the moment to remove the editor, but I really caution anyone actually using the provided editor as it seems to be being used as a magic bullet for solutions that are better served by many other patterns.

**Comment by lierdakil** [comment]
@park9140 Thank you for pointing out that thing about cahcing on buffer, I actually did not consider a case with multiple editors for the same buffer. I probably should fix that.

I'd love to see an interface for "prefix providers", as you put it, but while there is none, using editor instance to recalculate prefix is the best existing option, and I'd rather have it stay, at least until better alternatives are provided. And I'd prefer to have easy access to cursor that triggered autocompletion as well, pretty much for the same reason. We could discuss this further when there actually *is* a proposed interface for specifying custom prefixes.

**Comment by joefitzgerald** [comment]
Now this is interesting. Prefix providers. I like that concept.

**Comment by benogle** [comment]
I want to leave cursor out of the options.

Using `cursor` for prefix generation is a hack. The only argument for having cursor seems to be to use `Cursor::getBeginningOfCurrentWordBufferPosition`, which isnt a great proxy for what you _really_ want want. What you really want is a nice way to match a regex on a given range of the buffer. We should add a method to `editor`, and stop using the cursor altogether. Something like this is really what you want 

```coffee
# Naming could be better!!

# note the `^` and `$` operators would match the beginning and end of the passed in range.
prefixBufferRange = editor.getMatchBufferRange([[position.row, 0], position], /[\w]+$/)
prefix = editor.getTextInRange(prefixBufferRange)

# Or maybe with no positional information:
prefix = editor.matchForBufferRange([[position.row, 0], position], /[\w]+$/)
```

Until a method like this is available, here is a better way to generate the prefix than using the cursor methods:

```coffee
getPrefix = (editor, position) ->
  regex = /[\w]+$/ # whatever your prefix regex might be
  line = editor.getTextInRange([[position.row, 0], position])
  line.match(regex)?[0]
```

In your providers today, you can assume the cursor is the last cursor. It is as simple as `editor.getLastCursor()` if you really want to use it.

As for the other options, I would like to keep them as they are. `position` (which I will rename to `bufferPosition` for clarity) , `scopeDescriptor` are for convenience in the 80% case. You will often need to pass your buffer position to some executable (clang?), or generate the prefix via the `getPrefix` method defined above, or match scopes.

**Comment by lierdakil** [comment]
@benogle
First, ad-hoc extending public class instances is a path to disaster IMO, since they can (and likely eventually will) clash with prototype changes or other packages. Please consider providing such utility functions separately, e.g. as methods of options object (which would be confusing as hell, but safe -- there has to be a better solution that I don't see right now)

Second, cursor would be useful for things like lookaround and ad-hoc buffer preprocessing. Please understand that not every completion is defined only by immediately preceding text, and not every grammar is rich enough (or even could be). Granted, this could be done by clever use of regex and `Point` arithmetics. But using cursor could reduce boilerplate significantly.

I really don't think that cutting down API interface is the right way to go. If you want to discourage use of certain patterns, you could a) document it b) move accessors to, e.g. `extendedOptions`, if you really have to. As I said earlier, making developers, who want to do something you have not foreseen, jump through hoops, is not a great way to promote a platform. If it's supposed to be open platform, why close down API then?

Anyway, I think I made my point during the course of this discussion. Please think on my arguments. If you're still unconvinced, well, I don't think I can convince you then. Thank you for your time, I appreciate at least *having* a discussion :v:

**Comment by benogle** [comment]
> First, ad-hoc extending public class instances is a path to disaster IMO, since they can (and likely eventually will) clash with prototype changes or other packages. Please consider providing such utility functions separately, e.g. as methods of options object (which would be confusing as hell, but safe -- there has to be a better solution that I don't see right now)

I dont understand what you mean. I'm certainly not going to monkey patch editor, if that's what you're insinuating. It would be an addition to the public TextEditor API. Not right this second, but soon-ish. I think this prefix usecase is a good argument for a nice way to match a regex on a range in the buffer.

> Second, cursor would be useful for things like lookaround and ad-hoc buffer preprocessing. Please understand that not every completion is defined only by immediately preceding text, and not every grammar is rich enough (or even could be). Granted, this could be done by clever use of regex and Point arithmetics. But using cursor could reduce boilerplate significantly.

I do not see how the cursor provides extra help. Even if it does in some corner case, that doesnt mean that it needs to be included in every request. If you really really really need it, then you can get it from editor.

> If you want to discourage use of certain patterns, you could a) document it b) move accessors to, e.g. extendedOptions

You can already get nearly everything you want from editor.
